aws-storage-s3/src/androidTest/java/com/amplifyframework/storage/s3/AWSS3StorageUploadTest.java

```diff
@@ -0,0 +1,238 @@
+
+    private StorageUploadFileOptions options;
+
+    private Set<SubscriptionToken> subscriptions;
+
+    /**
+     * Create temp files to upload ahead of time.
+     * @throws Exception if file creation or upload fails
+     */
+    @BeforeClass
+    public static void setUpBeforeClass() throws Exception {
+        // Randomly write objects to upload
+        largeFile = createTempFile(LARGE_FILE_NAME, LARGE_FILE_SIZE);
+        smallFile = createTempFile(SMALL_FILE_NAME, SMALL_FILE_SIZE);
+    }
+
+    /**
+     * Cleans up the S3 bucket for files that were uploaded
+     * during testing processes.
+     */
+    @AfterClass
+    public static void cleanUp() {
+        String largeFileKey = getS3Key(DEFAULT_ACCESS_LEVEL, LARGE_FILE_NAME);
+        String smallFileKey = getS3Key(DEFAULT_ACCESS_LEVEL, SMALL_FILE_NAME);
+        cleanUpS3Object(largeFileKey);
+        cleanUpS3Object(smallFileKey);
+    }
+
+    /**
+     * Sets up the options to use for transfer.
+     */
+    @Before
+    public void setUp() {
+        // Always interact with PUBLIC access for consistency
+        options = StorageUploadFileOptions.builder()
+                .accessLevel(DEFAULT_ACCESS_LEVEL)
+                .build();
+
+        // Create a set to remember all the subscriptions
+        subscriptions = new HashSet<>();
+    }
+
+    /**
+     * Unsubscribe from everything after each test.
+     */
+    @After
+    public void unsubscribe() {
+        // Unsubscribe from everything
+        for (SubscriptionToken token : subscriptions) {
+            Amplify.Hub.unsubscribe(token);
+        }
+    }
+
+    /**
+     * Tests that small file (single-part) uploads successfully.
+     *
+     * @throws Exception if upload fails
+     */
+    @Test
+    public void testUploadSmallFile() throws Exception {
+        latchedUploadAndConfirm(smallFile, DEFAULT_ACCESS_LEVEL, getIdentityId());
+    }
+
+    /**
+     * Tests that large file (multi-part) uploads successfully.
+     *
+     * @throws Exception if upload fails
+     */
+    @Test
+    public void testUploadLargeFile() throws Exception {
+        latchedUploadAndConfirm(largeFile, DEFAULT_ACCESS_LEVEL, getIdentityId());
+    }
+
+    /**
+     * Tests that file upload operation can be canceled while the
+     * transfer hasn't completed yet.
+     *
+     * @throws Exception if upload is not canceled successfully
+     *         before timeout
+     */
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testUploadFileIsCancelable() throws Exception {
+        final CountDownLatch canceled = new CountDownLatch(1);
+
+        // Begin uploading large file
+        StorageUploadFileOperation<?> op = Amplify.Storage.uploadFile(
+            largeFile.getName(),
+            largeFile.getAbsolutePath(),
+            options,
+            onResult -> fail("Upload finished before being successfully cancelled."),
+            onError -> fail("Upload failed for a different reason.")
```

Comment:
These probably won't work the way you expect. They will crash the thread on which these callbacks occur, not the jUnit thread which is ready to intercept any thrown exceptions and report them. See notes about `Await` below.

Reply:
fixed